pkg/search/rules/insertrule.go

```diff
@@ -0,0 +1,86 @@
+package rules
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/mark3labs/mcp-go/server"
+
+	"github.com/algolia/algoliasearch-client-go/v3/algolia/search"
+	"github.com/algolia/mcp/pkg/mcputil"
+)
+
+const (
+	rulesBaseURL = "https://%s.algolia.net/1/indexes/%s/rules/%s"
+)
+
+func RegisterInsertRule(mcps *server.MCPServer, index *search.Index, appID, apiKey string) {
+	insertRuleTool := mcp.NewTool(
+		"save_rule",
+		mcp.WithDescription("Save or update a rule in the Algolia index"),
+		mcp.WithString(
+			"objectID",
+			mcp.Description("The unique identifier of the rule"),
+			mcp.Required(),
+		),
+		mcp.WithString(
+			"rule",
+			mcp.Description("The rule object as a JSON string. Example schema: {\"objectID\":\"pattern\",\"conditions\":[{\"anchoring\":\"contains\",\"pattern\":\"pattern\"}],\"consequence\":{\"params\":{\"filters\":\"attribute:value\",\"query\":\"\"},\"promote\":[{\"objectID\":\"id\",\"position\":0}]},\"description\":\"description\"}"),
+			mcp.Required(),
+		),
+	)
+
+	mcps.AddTool(insertRuleTool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		indexName := index.GetName()
+		objectID, ok := req.Params.Arguments["objectID"].(string)
+		if !ok {
+			return mcp.NewToolResultError("invalid objectID format"), nil
+		}
+
+		ruleStr, ok := req.Params.Arguments["rule"].(string)
+		if !ok {
+			return mcp.NewToolResultError("invalid rule format, expected JSON string"), nil
+		}
+
+		// Create HTTP client
```

Comment:
We should add this to the SDK if it doesn't exist.

Reply:
The SDK requires only allow to create rules and synonyms programatically, it's not possible to use a json format requiring to rebuild an interface (mcp input to api client input) that's why I used it.

I will check if the v4 of the client allows to improve it.

Reply:
I did the upgrade and simplified the add synonym tool: https://github.com/algolia/mcp/pull/29